Correction of Zero Solar Load in the Zone Component Load Summary Report

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #7608 
 - Corrects the zero solar load as well as other zero loads that show up in the delayed column.  The problem was that decay curves were not being calculated when the user selected design days from the weather file period.  This was due to the fact that the delay curves were based on factors that were only calculated on the first day of the simulation.  This changes things so that those factors are calculated on every day and thus will be available for whatever day is selected as the design day.  Values are now reported as non-zero when appropriate.

### Pull Request Author: Rick Strand
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correction of issue

This commit fixes the problem and the defect idf now shows solar output.

RKStrand committed:
Creation of unit test

An unit test for the bug fix was added.

JasonGlazer commented:
@RKStrand I am not sure I agree with your solution since the calculation of the decay curves is time consuming, recalculating them each day could substantially increase the calculation time for a simulation using them. Have you looked at the performance impact?

RKStrand commented:
@JasonGlazer I'll agree that the solution does not seem ideal but I wasn't sure there was another option.  When one is looking at a design day, it's pretty obvious that it's just the first day of the solution.  When the user asks for a design day from the weather file though, one doesn't know beforehand which day of the year that will be.  Or is there a way to figure that out before doing the simulation for that weather environment.  That's where I got stuck and felt like the only way to guarantee that it would do this was doing the calculation every day.  Maybe I missed something in the logic though and I'm happy to alter this if I misunderstood something.  I have not looked at the performance impact but when I looked at the code I thought that the number of calculations wasn't that large.  But perhaps I misunderstood that also?

JasonGlazer commented:
There are some nested loops of calculations so that is why a believe they are expensive. 

Maybe make the logic different if the user is using design days vs the weather year?

Even if using a weather day, if it is not setting the peak cooling or peak heating that day, then these calculations are not needed.

RKStrand committed:
Merge branch 'develop' into 7608ZeroSolarLoadZoneCompSummWeatherFileDays

RKStrand commented:
@JasonGlazer So, I have altered the logic of when to calculate the pulse.  Now, it only calculates it on the first day of the simulation (as before) OR when the user selects to run the weather file as the sizing period.  If the user has selected the weather file as the sizing period, then it will still calculate these every day because it seems to me that it does not calculate which day of the year should be the design day a priori but rather on the fly.  Is that a correct understanding of the code or am I missing something?

From what I have seen, it seems to indicate that it doesn't know whether or not a particular day is going to be the design day yielding maximum loads until it has actually done the day.  So, doesn't that mean it has to calculate these pulse factors every day just in case a particular day is the actual day for sizing?

With the change I am proposing (haven't put it on GitHub yet but will soon), it only does the calculations every day when the user wants to use the entire year as the design day.  By doing this, they are already willing to wait longer for results.  So, I feel like a slight performance penalty is probably not a huge deal here, and I'm not sure there is any way around that.

What are your thoughts on this strategy?

JasonGlazer commented:
Rick, that sounds reasonable to me.

RKStrand commented:
Thanks, @JasonGlazer.  I am currently making sure that energyplus_test runs with the modified subroutine.  Once it does, I will put the new stuff in GitHub and probably also merge in develop.  Hopefully that will happen before lunch today.

RKStrand committed:
Correction to address review comment

The method used to decide when the component loads pulse happens was changed as a result of the initial review.  Now, it will only do the pulse on the first day of a particular simulation or on all days when the user has selected to have a weather file serve as the design environment.

RKStrand committed:
Merge branch 'develop' into 7608ZeroSolarLoadZoneCompSummWeatherFileDays

JasonGlazer commented:
@RKStrand any idea what the diffs are about from the W64 CI machine? I would not think it would impact convergence since it is a side calc.

mjwitte commented:
@JasonGlazer Those diffs are likely due to develop moving before Win64 CI got around to running this (2 days after the other platforms). I would ignore those. If you want to confirm, you could run one of the affected files locally with develop and this branch.

RKStrand commented:
@mjwitte Thanks for the clarification, Mike.  Hopefully that explains the diffs.

JasonGlazer committed:
Merge branch 'develop' into 7608ZeroSolarLoadZoneCompSummWeatherFileDays

JasonGlazer commented:
Code review, reran unit tests locally, reran original issue. Looks good. The CI results no longer present so pushed merged version up to trigger again. At this point, I would say it is good to go unless performance is somehow impacted or CI is not green.

Myoldmopar commented on src/EnergyPlus/SizingManager.cc:
```diff
@@ -844,6 +833,31 @@ namespace SizingManager {
+        }
+        
+    }
+
```
Wrapping this in a small, well-defined function is good.  And giving those magic numbers meaningful names is also a good thing.

Myoldmopar commented on tst/EnergyPlus/unit/SizingManager.unit.cc:
```diff
@@ -390,3 +390,98 @@ TEST_F(EnergyPlusFixture, SizingManager_DOASControlStrategyDefaultSpecificationT
     ASSERT_EQ(23.9, ZoneSizingInput(1).DOASHighSetpoint);
 }
+
+TEST_F(EnergyPlusFixture, SizingManager_CalcdoLoadComponentPulseNowTest)
```
Very nice unit test.  :+1:

Myoldmopar reviewed (approved):
All good from me.  I concur with @JasonGlazer that if CI comes back clean, this can merge.

Myoldmopar commented:
EDD diffs are unrelated.  I'm ready for this to go in.  Thanks @RKStrand for the changes and @JasonGlazer for the review.
